### PR TITLE
drivers: hwinfo: Add shell command for hwinfo device id

### PR DESCRIPTION
**Pull request by alexanderwachter**
This commit adds support for device id shell command.
Example:
uart:~$ hwinfo devid
Length: 12
ID: 0x1b0320d51485330313420

Signed-off-by: Alexander Wachter <alexander.wachter@student.tugraz.at>

### CODE REVIEW

File: drivers/hwinfo/hwinfo_shell.c [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+#include <misc/printk.h>
+#include <shell/shell.h>
+#include <hwinfo.h>
+#include <zephyr/types.h>
+#include <logging/log.h>
+
+#define LOG_LEVEL CONFIG_LOG_DEFAULT_LEVEL
+
+LOG_MODULE_REGISTER(hwinfo_shell);
```
Comment (carlescufi) [inline drivers/hwinfo/hwinfo_shell.c]: do you need a logger module here?

Comment (alexanderwachter) [reply drivers/hwinfo/hwinfo_shell.c]: No. I copied that from gpio shell.

Comment (alexanderwachter) [reply drivers/hwinfo/hwinfo_shell.c]: Removed.

File: drivers/hwinfo/hwinfo_shell.c [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+static int cmd_get_device_id(const struct shell *shell, size_t argc, char **argv)
+{
+	u8_t dev_id[16];
+	ssize_t length;
+	int i;
+
+	length = hwinfo_get_device_id(dev_id, sizeof(dev_id));
+
+	if (length == -ENOTSUP) {
+		shell_fprintf(shell, SHELL_ERROR, "Not supported by hardware\n");
```
Comment (jarz-nordic) [inline drivers/hwinfo/hwinfo_shell.c]: You can use macro `shell_error`:
`shell_error(shell, "Not supported by hardware");`

File: drivers/hwinfo/hwinfo_shell.c [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+		return 0;
+	} else if (length < 0) {
+		shell_fprintf(shell, SHELL_ERROR, "Error: %d\n", length);
```
Comment (jarz-nordic) [inline drivers/hwinfo/hwinfo_shell.c]: You can use `shell_error`.

File: drivers/hwinfo/hwinfo_shell.c [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+{
+	u8_t dev_id[16];
+	ssize_t length;
+	int i;
+
+	length = hwinfo_get_device_id(dev_id, sizeof(dev_id));
+
+	if (length == -ENOTSUP) {
+		shell_fprintf(shell, SHELL_ERROR, "Not supported by hardware\n");
+		return 0;
```
Comment (jarz-nordic) [inline drivers/hwinfo/hwinfo_shell.c]: When there is an error you shall return some error code. 0 means success execution.

File: drivers/hwinfo/hwinfo_shell.c [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+	} else if (length < 0) {
+		shell_fprintf(shell, SHELL_ERROR, "Error: %d\n", length);
+		return 0;
```
Comment (jarz-nordic) [inline drivers/hwinfo/hwinfo_shell.c]: When there is an error you shall return some error code. 0 means success execution.

File: drivers/hwinfo/hwinfo_shell.c [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+	}
+
+	shell_fprintf(shell, SHELL_NORMAL, "Length: %d\n", length);
```
Comment (jarz-nordic) [inline drivers/hwinfo/hwinfo_shell.c]: You can use macro: `shell_print`

`shell_print(shell, "Length: %d", length);`

File: drivers/hwinfo/hwinfo_shell.c [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+	return 0;
+}
+
+SHELL_CREATE_STATIC_SUBCMD_SET(sub_hwinfo) {
+	/* Alphabetically sorted. */
+	SHELL_CMD(devid, NULL, "Show device id", cmd_get_device_id),
+	SHELL_SUBCMD_SET_END /* Array terminated. */
+};
+
+SHELL_CMD_REGISTER(hwinfo, &sub_hwinfo, "HWINFO commands", NULL);
```
Comment (jarz-nordic) [inline drivers/hwinfo/hwinfo_shell.c]: You can use following macro:
`SHELL_CMD_ARG_REGISTER(hwinfo, &sub_hwinfo, "HWINFO commands", NULL, 2, 0);`

File: samples/subsys/shell/shell_module/prj.conf [hunk 8]
```diff
@@ -8,3 +8,4 @@ CONFIG_BOOT_BANNER=n
 CONFIG_THREAD_NAME=y
 CONFIG_DEVICE_SHELL=y
 CONFIG_GPIO_SHELL=y
+CONFIG_HWINFO_SHELL=y
```
Comment (nashif) [inline samples/subsys/shell/shell_module/prj.conf]: I think you would want CONFIG_HWINFO=y here as well. Maybe put this in tests/boards/board_shell/, it more appropriate than the shell sample.

File: drivers/hwinfo/hwinfo_shell.c [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+		shell_fprintf(shell, SHELL_NORMAL, "%x", dev_id[i]);
+	}
+
+	shell_fprintf(shell, SHELL_NORMAL, "\n");
+	return 0;
+}
+
+SHELL_CREATE_STATIC_SUBCMD_SET(sub_hwinfo) {
+	/* Alphabetically sorted. */
+	SHELL_CMD(devid, NULL, "Show device id", cmd_get_device_id),
```
Comment (jarz-nordic) [inline drivers/hwinfo/hwinfo_shell.c]: `SHELL_CMD_ARG(devid, NULL, "Show device id", cmd_get_device_id, 1, 0),`

**Review by jakub-uC — COMMENTED** [review]
Please update my minor proposals.

File: samples/subsys/shell/shell_module/prj.conf (hunk 8 above)
Comment (alexanderwachter) [reply samples/subsys/shell/shell_module/prj.conf]: I thought the shell sample should have all shell command built in.

File: drivers/hwinfo/hwinfo_shell.c (hunk 6 above)
Comment (alexanderwachter) [reply drivers/hwinfo/hwinfo_shell.c]: @jarz-nordic the problem with shell_print is the "\n" that is added at the end of the string

File: drivers/hwinfo/hwinfo_shell.c (hunk 2 above)
Comment (alexanderwachter) [reply drivers/hwinfo/hwinfo_shell.c]: Done.

File: drivers/hwinfo/hwinfo_shell.c (hunk 3 above)
Comment (alexanderwachter) [reply drivers/hwinfo/hwinfo_shell.c]: Done.

File: drivers/hwinfo/hwinfo_shell.c (hunk 4 above)
Comment (alexanderwachter) [reply drivers/hwinfo/hwinfo_shell.c]: Done.

File: drivers/hwinfo/hwinfo_shell.c (hunk 5 above)
Comment (alexanderwachter) [reply drivers/hwinfo/hwinfo_shell.c]: Done.

File: drivers/hwinfo/hwinfo_shell.c (hunk 7 above)
Comment (alexanderwachter) [reply drivers/hwinfo/hwinfo_shell.c]: Done.

File: drivers/hwinfo/hwinfo_shell.c (hunk 9 above)
Comment (alexanderwachter) [reply drivers/hwinfo/hwinfo_shell.c]: Done.

File: samples/subsys/shell/shell_module/prj.conf (hunk 8 above)
Comment (alexanderwachter) [reply samples/subsys/shell/shell_module/prj.conf]: Should i remove it?
Added it to tests/boards/board_shell

**Review by jakub-uC — APPROVED** [review]

File: samples/subsys/shell/shell_module/prj.conf (hunk 8 above)
Comment (nashif) [reply samples/subsys/shell/shell_module/prj.conf]: lets keep the shell sample dedicated to shell features, I will need probably to move all the HW related shell to the board_shell instead.

**Review by nashif — APPROVED** [review]

**Pull request merged**